### PR TITLE
Python: add ZRANDMEMBER command

### PR DESCRIPTION
**Pull request by aaron-congo**

### CODE REVIEW

File: python/python/glide/async_commands/core.py [hunk 1]
```diff
@@ -2746,6 +2746,89 @@ async def zdiffstore(self, destination: str, keys: List[str]) -> int:
...
+                "GLIDE"  # "GLIDE" is a random member of "my_sorted_set".
+            >>> await client.zrandmember("non_existing_sorted_set")
+                None  # "non_existing_sorted_set" is not an existing key, so None was returned.
+        """
+        return cast(
+            Optional[str],
+            await self._execute_command(RequestType.ZRandMember, [key]),
+        )
+
+    async def zrandmember_count(self, key: str, count: int) -> List[str]:
```
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/core.py:2773]: Maybe `zrandmember_with_count` or `zrandmember_withcount`?

File: python/python/glide/async_commands/core.py [hunk 2]
```diff
@@ -2746,6 +2746,89 @@ async def zdiffstore(self, destination: str, keys: List[str]) -> int:
...
+                If negative, allows for duplicates.
+
+        Returns:
+            List[List[Union[str, float]]]: A list of `[member, score]` lists, where `member` is a random member from
+                the sorted set and `score` is the associated score.
+                If the sorted set does not exist or is empty, the response will be an empty list.
+
+        Examples:
+            >>> await client.zrandmember_withscores("my_sorted_set", -3)
+                [{"GLIDE": 1.0, "GLIDE": 1.0, "PYTHON": 2.0}]  # "GLIDE" and "PYTHON" are random members of "my_sorted_set", and have scores of 1.0 and 2.0, respectively.
```
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/core.py:2821]:
```suggestion
                [["GLIDE": 1.0], ["GLIDE": 1.0], ["PYTHON": 2.0]]  # "GLIDE" and "PYTHON" are random members of "my_sorted_set", and have scores of 1.0 and 2.0, respectively.
```

File: python/python/glide/async_commands/core.py [hunk 3]
```diff
@@ -2746,6 +2746,89 @@ async def zdiffstore(self, destination: str, keys: List[str]) -> int:
...
+            >>> await client.zrandmember_withscores("non_existing_sorted_set", 3)
+                []  # "non_existing_sorted_set" is not an existing key, so an empty list was returned.
+        """
+        return cast(
+            List[List[Union[str, float]]],
+            await self._execute_command(
+                RequestType.ZRandMember, [key, str(count), "WITHSCORES"]
```
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/core.py:2828]: Do we have a place to store magic constants in python client?

File: python/python/tests/test_async_client.py [hunk 4]
```diff
@@ -2378,6 +2378,84 @@ async def test_zdiffstore(self, redis_client: TRedisClient):
...
+    async def test_zrandmember_withscores(self, redis_client: TRedisClient):
+        key = get_random_string(10)
+        string_key = get_random_string(10)
+        scores = {"one": 1, "two": 2}
+        assert await redis_client.zadd(key, scores) == 2
+
+        # unique values are expected as count is positive
+        elements = await redis_client.zrandmember_withscores(key, 4)
+        assert len(elements) == 2
+        print("The score type: " + type(elements[0][1]).__name__)
```
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:2436]: remove

File: glide-core/src/client/value_conversion.rs [hunk 5]
```diff
@@ -287,6 +288,46 @@ pub(crate) fn convert_to_expected_type(
...
+                Value::Array(ref array)
+                    if array.is_empty() || matches!(array[0], Value::Array(_)) =>
+                {
+                    // The server response is an empty array or a RESP3 array of pairs, with scores of type double,
+                    // so it is already in the correct format.
+                    Ok(value)
+                }
+                Value::Array(mut array)
+                    if array.len() > 1
+                        && matches!(array[1], Value::BulkString(_) | Value::SimpleString(_)) =>
```
Comment (acarbonetto) [inline glide-core/src/client/value_conversion.rs:310]: should we also check that this array is of even size?

File: glide-core/src/client/value_conversion.rs [hunk 6]
```diff
@@ -287,6 +288,46 @@ pub(crate) fn convert_to_expected_type(
             )
                 .into()),
         },
+        // Used by ZRANDMEMBER when the WITHSCORES arg is passed.
+        // The server response can be a nil value, an empty array, a flat array of member-score pairs, or a two-dimensional array of member-score pairs.
+        // The server response scores can be strings or doubles. The conversions we do here are as follows:
+        //
+        // - if the server returned nil, return nil
+        // - if the server returned an empty array, return an empty array
```
Comment (acarbonetto) [inline glide-core/src/client/value_conversion.rs:296]: does this ever happen? I think the commands will return nil, and we won't ever get an empty array

File: python/python/glide/async_commands/core.py [hunk 7]
```diff
@@ -2746,6 +2746,89 @@ async def zdiffstore(self, destination: str, keys: List[str]) -> int:
...
+            key (str): The key of the sorted set.
+
+        Returns:
+            Optional[str]: A string representing a random element from the sorted set.
+                If the sorted set does not exist or is empty, the response will be None.
+
+        Examples:
+            >>> await client.zrandmember("my_sorted_set")
+                "GLIDE"  # "GLIDE" is a random member of "my_sorted_set".
+            >>> await client.zrandmember("non_existing_sorted_set")
```
Comment (acarbonetto) [inline python/python/glide/async_commands/core.py:2765]: we shouldn't include error cases - examples should demonstrate how to use the command, not how to not use it...

File: python/python/glide/async_commands/core.py [hunk 8]
```diff
@@ -2746,6 +2746,89 @@ async def zdiffstore(self, destination: str, keys: List[str]) -> int:
...
+        """
+        Retrieves random elements from the sorted set stored at 'key'.
+
+        See https://valkey.io/commands/zrandmember for more details.
+
+        Args:
+            key (str): The key of the sorted set.
+            count (int): The number of elements to return.
+                If `count` is positive, returns unique elements.
+                If negative, allows for duplicates.
```
Comment (acarbonetto) [inline python/python/glide/async_commands/core.py:2783]:
```suggestion
                If `count` is negative, allows for duplicate elements.
```

File: python/python/glide/async_commands/core.py [hunk 9]
```diff
@@ -2746,6 +2746,89 @@ async def zdiffstore(self, destination: str, keys: List[str]) -> int:
...
+
+        Returns:
+            List[str]: A list of elements from the sorted set.
+                If the sorted set does not exist or is empty, the response will be an empty list.
+
+        Examples:
+            >>> await client.zrandmember("my_sorted_set", -3)
+                ["GLIDE", "GLIDE", "PYTHON"]  # "GLIDE" and "PYTHON" are random members of "my_sorted_set".
+            >>> await client.zrandmember("non_existing_sorted_set", 3)
```
Comment (acarbonetto) [inline python/python/glide/async_commands/core.py:2792]: we should remove this example

File: python/python/glide/async_commands/core.py [hunk 10]
```diff
@@ -2746,6 +2746,89 @@ async def zdiffstore(self, destination: str, keys: List[str]) -> int:
...
+        """
+        Retrieves random elements along with their scores from the sorted set stored at 'key'.
+
+        See https://valkey.io/commands/zrandmember for more details.
+
+        Args:
+            key (str): The key of the sorted set.
+            count (int): The number of elements to return.
+                If `count` is positive, returns unique elements.
+                If negative, allows for duplicates.
```
Comment (acarbonetto) [inline python/python/glide/async_commands/core.py:2812]:
```suggestion
                If `count` is negative, allows for duplicate elements.
```

**Review by acarbonetto — APPROVED** [review]

File: python/python/glide/async_commands/core.py (hunk 1 above)
Comment (aaron-congo) [reply python/python/glide/async_commands/core.py:2773]: the other pre-existing functions in this file do not include a "with", for example spop_count, lpop_count etc

File: python/python/glide/async_commands/core.py (hunk 3 above)
Comment (aaron-congo) [reply python/python/glide/async_commands/core.py:2828]: We brought up the idea to create a constant for this but Sho said not to, see [here](https://github.com/aws/glide-for-redis/pull/1334#discussion_r1584425517)

File: glide-core/src/client/value_conversion.rs (hunk 5 above)
Comment (aaron-congo) [reply glide-core/src/client/value_conversion.rs:310]: Sure, I can add that

File: glide-core/src/client/value_conversion.rs (hunk 6 above)
Comment (aaron-congo) [reply glide-core/src/client/value_conversion.rs:296]: From the redis docs: "Array reply: when the additional count argument is passed, the command returns an array of members, or an empty array when key doesn't exist."

File: python/python/glide/async_commands/core.py (hunk 8 above)
Comment (aaron-congo) [reply python/python/glide/async_commands/core.py:2783]: I'll change this here and for hrandfield which has the same line. Note that the Java docs worded this the same so we may want to change it there as well

File: python/python/glide/async_commands/core.py (hunk 9 above)
Comment (aaron-congo) [reply python/python/glide/async_commands/core.py:2792]: See my comment [above](https://github.com/Bit-Quill/glide-for-redis/pull/276#discussion_r1600345031)

File: glide-core/src/client/value_conversion.rs [hunk 11]
```diff
@@ -287,6 +288,46 @@ pub(crate) fn convert_to_expected_type(
...
+        // - otherwise, return a two-dimensional array of member-score pairs, where scores are of type double
+        ExpectedReturnType::ArrayOfMemberScorePairs => {
```
[the remark this answers is not in the document]
Comment (aaron-congo) [reply glide-core/src/client/value_conversion.rs:289]: Updated now

**Review by Yury-Fridlyand — APPROVED** [review]

**Pull request merged**